### PR TITLE
test(callbacks): simplified mock contract keeper's processCallback logic

### PR DESCRIPTION
**Pull request by srdtrk**
## Description

This is one of the issues Colin requested be addressed sooner rather than later. Mock contract keeper's `processCallback` now relies on the contract address to take different actions rather than gas remaining. 

closes: https://github.com/cosmos/ibc-go/issues/4309

### Commit Message / Changelog Entry

```text
test(callbacks): simplified mock contract keeper's processCallback logic
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [x] ~Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).~
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [x] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/callbacks/testing/simapp/contract_keeper.go [hunk 1]
```diff
@@ -145,16 +162,25 @@ func (k ContractKeeper) processMockCallback(
...
-	}
-
-	if authAddress == MockCallbackUnauthorizedAddress {
-		ctx.GasMeter().ConsumeGas(500000, fmt.Sprintf("mock %s callback unauthorized", callbackType))
+	switch contractAddress {
+	case ErrorContract:
+		ctx.GasMeter().ConsumeGas(gasRemaining/2, fmt.Sprintf("mock %s callback unauthorized", callbackType))
 		return ibcmock.MockApplicationCallbackError
+	case OogPanicContract:
+		ctx.GasMeter().ConsumeGas(gasRemaining+1, fmt.Sprintf("mock %s callback oog panic", callbackType))
```
Comment (damiannolan) [inline modules/apps/callbacks/testing/simapp/contract_keeper.go:170]: gas remaining + 1 triggers the panic, right?

File: modules/apps/callbacks/testing/simapp/contract_keeper.go [hunk 2]
```diff
@@ -145,16 +162,25 @@ func (k ContractKeeper) processMockCallback(
...
-	if gasRemaining < 500000 {
-		// consume gas will panic since we attempt to consume 500_000 gas, for tests
-		ctx.GasMeter().ConsumeGas(500000, fmt.Sprintf("mock %s callback panic", callbackType))
-	}
-
-	if authAddress == MockCallbackUnauthorizedAddress {
-		ctx.GasMeter().ConsumeGas(500000, fmt.Sprintf("mock %s callback unauthorized", callbackType))
+	switch contractAddress {
+	case ErrorContract:
+		ctx.GasMeter().ConsumeGas(gasRemaining/2, fmt.Sprintf("mock %s callback unauthorized", callbackType))
```
Comment (damiannolan) [inline modules/apps/callbacks/testing/simapp/contract_keeper.go:167]: How come it needs to be gas remaining / 2? or does this have any implications?

**Review by damiannolan — APPROVED** [review]
Nice refactor! I'm happy enough with this approach if you are, would be good to get a couple more reviews :)

File: modules/apps/callbacks/testing/simapp/contract_keeper.go (hunk 1 above)
Comment (srdtrk) [reply modules/apps/callbacks/testing/simapp/contract_keeper.go:170]: right

File: modules/apps/callbacks/testing/simapp/contract_keeper.go (hunk 2 above)
Comment (srdtrk) [reply modules/apps/callbacks/testing/simapp/contract_keeper.go:167]: I just wanted an amount that wouldn't panic

Comment (colin-axner) [reply modules/apps/callbacks/testing/simapp/contract_keeper.go:167]: should we add this reasoning in the godoc?

File: modules/apps/callbacks/ibc_middleware_test.go [hunk 3]
```diff
@@ -272,10 +276,10 @@ func (s *CallbacksTestSuite) TestOnAcknowledgementPacket() {
 			s.SetupTransferTest()
 
 			// set user gas limit above panic level in mock contract keeper
```
Comment (colin-axner) [inline modules/apps/callbacks/ibc_middleware_test.go:278]: I think this comment is unnecessary now

File: modules/apps/callbacks/ibc_middleware_test.go [hunk 4]
```diff
@@ -272,10 +276,10 @@ func (s *CallbacksTestSuite) TestOnAcknowledgementPacket() {
...
-			userGasLimit := 600000
+			userGasLimit = 600000
```
Comment (colin-axner) [inline modules/apps/callbacks/ibc_middleware_test.go:279]: does this get modified? Could probably be defined at the top of the test

File: modules/apps/callbacks/ibc_middleware_test.go [hunk 5]
```diff
@@ -578,7 +585,7 @@ func (s *CallbacksTestSuite) TestOnRecvPacket() {
 			s.SetupTransferTest()
 
 			// set user gas limit above panic level in mock contract keeper
-			userGasLimit := 600_000
+			userGasLimit = 600_000
```
Comment (colin-axner) [inline modules/apps/callbacks/ibc_middleware_test.go:588]: ditto

**Review by colin-axner — APPROVED** [review]
This is great! Thank you @srdtrk! I think this will be easier to maintain in the long term when you need to revisit the testing setup after some time passes

File: modules/apps/callbacks/ibc_middleware_test.go (hunk 4 above)
Comment (srdtrk) [reply modules/apps/callbacks/ibc_middleware_test.go:279]: No, but the malleate needs access to this variable if it needs to change the memo. If you define it at the top, then an unexpected behavior can occur in which the modification of this variable from a test case can leak to the next test case

File: modules/apps/callbacks/ibc_middleware_test.go (hunk 3 above)
Comment (srdtrk) [reply modules/apps/callbacks/ibc_middleware_test.go]: done. Deleted the comment

File: modules/apps/callbacks/ibc_middleware_test.go (hunk 5 above)
Comment (srdtrk) [reply modules/apps/callbacks/ibc_middleware_test.go:587]: :point_up: 

> No, but the malleate needs access to this variable if it needs to change the memo. If you define it at the top, then an unexpected behavior can occur in which the modification of this variable from a test case can leak to the next test case

**Pull request merged**